lib/avalon/batch/entry.rb

```diff
@@ -277,15 +277,32 @@ def self.caption_language(language)
       end
       private_class_method :caption_language
 
+      def self.process_datastream(datastream, type)
+        file, label, language = ["_file", "_label", "_language"].map { |item| item.prepend(type).to_sym }
+        if datastream[file].present? && FileLocator.new(datastream[file]).exist?
+          # Build out file metadata
+          filename = datastream[file].split('/').last
+          label = datastream[label].presence || filename
+          language = datastream[language].present? ? caption_language(datastream[language]) : Settings.caption_default.language
+          machine_generated = datastream[:machine_generated].present? ? 'machine_generated' : nil
+          # Create SupplementalFile
+          supplemental_file = SupplementalFile.new(label: label, tags: [type, machine_generated].compact, language: language)
+          supplemental_file.file.attach(io: FileLocator.new(datastream[file]).reader, filename: filename)
+          supplemental_file.save
```

Comment:
What should happen if this save fails?  Using `save!` would raise an error.  Alternatively this method could return nil if the save failed.  Or is there a place where errors on the unsaved `supplemental_file` object would get read and returned to the user?  I think the last might be best but maybe returning nil and handling that in `attach_datastreams_to_master_file` might be easiest for now.  What do you think?

Reply:
That's a good concern. I would think overall behavior would be that the batch ingest completes and information is provided in the completion email about how the media items were created but there were problems with X supplemental file(s)? 

So maybe switch over to `save!` and create a catch that adds to the email? Or would having error information in a "Batch processed successfully" email be confusing? I think there are currently 2 emails that get sent: one about the batch being valid and then one about processing being successful or failing. Maybe we add a third set of emails that deals specifically with the caption/transcript files?